SONARPY-2861: Create rule S7496: Creation of collections with literals and comprehensions should not be wrapped in type constructors (#5023)
SonarSource/rspec: rules/S7496/python/rule.adoc

This rule raises an issue when calls to `tuple()`, `list()`, `set()`, or `dict()` 

> are done with an argument that is already a collection literal or a comprehension that produces either the target type or an intermediate type. 

Reviewer: this sounds really complicated, but I don't really have a better idea from the top of my had either. Maybe we can come up with something together

Reviewer 2: I went with 
```
This rule raises an issue when calls to `tuple()`, `list()`, `set()`, or `dict()` 
are done with an argument that is either already a collection literal or a comprehension.
```
because the type is always either a type or not the type, which would be intermediate.

Reviewer: This is way better

== Why is this an issue?

Python provides direct ways to create common data structures like tuples, lists, and dictionaries using literals (e.g., `()`, `(1,2)`, `[]`, `[1,2]`, `{}`, `{'a':1}`) 

> and comprehensions (e.g., `[x for x in y]`, `{k:v for k,v in y}`).

Reviewer: Parenthesis should be avoided according to the style guide

> Wrapping these direct forms in a type constructors, for example:

Reviewer: It feels a bit clunky that the code block interrupts the sentence

```suggestion
Wrapping these direct forms in a type constructors is unnecessary, as can be seen in the following example:
```

Reviewer 2: Yes that's better

Reviewer: This sounds way better

[source, python]
----
tuple((1,2))

list([1,2])
list([x for x in [1,2]])

set({})
set({for k in [1,2]})

dict({'a':1})
dict({k:v for k,v in [1,2]})
dict([(1,"a"), (2, "b")]) 
----

> is unnecessary. 

Reviewer: ```suggestion
```

Such constructs:

> * add overhead by creating an intermediate collections

Reviewer: missing empty line for asciidoc to format the list properly
```suggestion

* add overhead by creating an intermediate collections
```

* add verbosity without providing any additional functionality
* add ambiguity and may mislead readers or imply a more complex operation than what is actually occurring

=== Exceptions

> When using comprehensions this rule will only raise issues when the comprehension filter or modify the elements.

Reviewer: Do you think it would be worth mentioning the reason for this?

== How to fix it

> To fix this remove the constructor call and use the literal or comprehension syntax of the target type.

Reviewer: ```suggestion
To fix this, remove the constructor call and use the literal or comprehension syntax of the target type.
```

As an exception, for a list or comprehension inside of a `tuple` constructor replace the comprehension by a generator.

[source,python,diff-id=1,diff-type=noncompliant]
----
t1 = tuple([1, 2]) # Noncompliant: the tuple creation can be done directly without the intermediate list
----

[source,python,diff-id=1,diff-type=compliant]
----
t1 = (1, 2) # Compliant
----

[source,python,diff-id=2,diff-type=noncompliant]
----
def foo(x):
    return x * 2

l1 = list([foo(x) for x in [1,2,3]]) # Noncompliant: list is redundant.
----

[source,python,diff-id=2,diff-type=compliant]
----
def foo(x):
    return x * 2

> l1 = [foo(x) for x in [1,2,3]] # Compliant

Reviewer: the space after the foo function isn't consistent.

Also, the function could be inlined: `[x * 2 for x in [1, 2, 3]]`

Reviewer 2: Yes inline is better

----

[source,python,diff-id=3,diff-type=noncompliant]
----
def foo(x):
    return x * 2

s1 = set([foo(x) for x in [1,2,3]]) # Noncompliant: the set creation can be done with a set comprehension.
----

[source,python,diff-id=3,diff-type=compliant]
----
def foo(x):
    return x * 2
s1 = {foo(x) for x in [1,2,3]} # Compliant
----

== Resources

=== Documentation
* Python Documentation - https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions[List Comprehensions]
* Python Documentation - https://docs.python.org/3/tutorial/datastructures.html#dictionaries[Dictionaries]
* Python Documentation - https://docs.python.org/3/tutorial/datastructures.html#sets[Sets]
* Python Documentation - https://docs.python.org/3/tutorial/datastructures.html#tuples-and-sequences[Tuples and Sequences]

ifdef::env-github,rspecator-view[]
== Implementation Specification

> 

Reviewer: It might be worth highlighting how to check if a comprehension does nothing, since there are some pitfalls there.

No filtering, no transformation, no multiple `for`s, and, in case of a dict comprehension, no flipping key and value.

=== Message

> When the inner type is the same as the constructor:
> * Remove the redundant {tuple|list|set|dict} constructor call.

Reviewer: ```suggestion
When the inner type is the same as the constructor:

* Remove the redundant {tuple|list|set|dict} constructor call.
```

When the inner type is different than the constructor:

> For literals:
> * Replace this {tuple|list|set|dict} constructor call by a {tuple|list|set|dict} literal.
> For comprehensions:
> * Replace this {list|set|dict} constructor call by a {list|set|dict} literal.
> 
> Exception for tuple with a list/set-comprehension `tuple([foo(x) in [1,2]])` 
> * Replace this list/set comprehension by a generator.

Reviewer: ```suggestion
For literals:

* Replace this {tuple|list|set|dict} constructor call by a {tuple|list|set|dict} literal.

For comprehensions:

* Replace this {list|set|dict} constructor call by a {list|set|dict} literal.

Exception for tuple with a list/set-comprehension `tuple([foo(x) in [1,2]])` 

* Replace this list/set comprehension by a generator.
```
